### PR TITLE
WIP for Seeing Event History

### PR DESCRIPTION
**Pull request by lukko**
First simple setup of the `historycontroller`.

Somehow I get the error as commented in `historycontroller.js` and I can't figure out how to solve it.

### CODE REVIEW

**Comment by N-Wouda** [comment]
You're using `this` in a static context.

File: index.html [hunk 1]
```diff
@@ -11,6 +11,14 @@
             <section id="factory" class="col-lg-4"></section>
             <section id="warehouse" class="col-lg-4"></section>
             <section id="customers" class="col-lg-4"></section>
+
+            <section id="history" class="col-lg-4">
```
Comment (N-Wouda) [inline index.html]: Please move the section's contents away into a different template.

File: src/controllers/historycontroller.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
+class HistoryController extends Controller
+{
+    static addEvent(event)
```
Comment (N-Wouda) [inline src/controllers/historycontroller.js]: Does this have to be a static method? I wrote a `registerEvent` method in the parent class to register jQuery event bindings, but I don't think what you are trying to achieve here involves events so much.

Perhaps just keep them as separate non-static functions?

**Comment by lukko** [comment]
What should I change it to?

File: index.html (hunk 1 above)
Comment (lukko) [reply index.html]: I know, it's just for testing.

File: src/controllers/historycontroller.js (hunk 2 above)
Comment (lukko) [reply src/controllers/historycontroller.js]: No, has nothing to do with jQuery events.
I thought it was kind of logical to use `HistoryController.addEvent(event)` in any other controller (you do not need to make a `new HistoryController` first), and pass any `Customer/Container/FactoryOrder` with it, to store it in the `history` array.

**Comment by lukko** [comment]
What's the `this` equivalent in static methods?

File: src/controllers/historycontroller.js (hunk 2 above)
Comment (N-Wouda) [reply src/controllers/historycontroller.js]: Perhaps turn this into a generic `log(entry)` method. No need for it to be static, and please don't use the word `event` outside the jQuery/event-binding context :).

**Comment by N-Wouda** [comment]
> What's the this equivalent in static methods?

No such thing exists, by design. I don't think your logging method needs to be static, so a 'normal' method would suffice.

**Comment by lukko** [comment]
What do you think of this?

**Comment by N-Wouda** [comment]
Will look at this tomorrow!

File: index.html [hunk 3]
```diff
@@ -11,14 +11,7 @@
...
             <section id="customers" class="col-lg-4"></section>
-
-            <section id="history" class="col-lg-4">
-                <div class="panel panel-default">
-                    <div class="panel-heading">History</div>
-                    <ul class="list-group" id="history-events"></ul>
-                </div>
-            </section>
-            
+            <section id="history" class="col-lg-4"></section>
```
Comment (N-Wouda) [inline index.html]: 👍

File: src/controllers/historycontroller.js [hunk 4]
```diff
@@ -1,47 +1,46 @@
...
-             * The above method calls are not working,
-             * I get this error:
-             *
-             * Uncaught TypeError: this._loadTemplate is not
-             * a function at Function.addEvent
-             * (http://localhost:8888/Serious-Game-for-Kids/src/controllers/historycontroller.js:9:18)at <anonymous>:1:19
-             */
+    log(entry)
+    {
+        if (entry instanceof Customer)
```
Comment (N-Wouda) [inline src/controllers/historycontroller.js]: I do not think this warrants two different methods. You're essentially doing the same thing in both, just using different templates. I'd refactor `addContainer()` and `addCustomer()` into `log(entry)`, as they are not too dissimilar (i.e. specific).

File: src/views/template/history.html [hunk 5]
```diff
@@ -0,0 +1,6 @@
+<div class="panel panel-default">
+    <div class="panel-heading">
+        History
+    </div>
+    <ul class="list-group" id="history-events"></ul>
+</div>
```
Comment (N-Wouda) [inline src/views/template/history.html]: Single empty lines please.

File: src/views/template/history.html [hunk 6]
```diff
@@ -0,0 +1,6 @@
+<div class="panel panel-default">
+    <div class="panel-heading">
+        History
```
Comment (N-Wouda) [inline src/views/template/history.html]: Wrap text in `{{#l}}History{{/l}}`, and add `History` to the `nl.json`, with a translation. Same thing goes for the other non-translated strings/variables.

**Comment by lukko** [comment]
I think this closes #22.
Might need to add states to order the `Order` model, to determine the right Bootstrap classes. But that would be better to have its own PR.

File: src/controllers/core/controller.js [hunk 7]
```diff
@@ -59,6 +59,8 @@ class Controller
 
                 if (append) {
                     $(anchor).append(template);
+                } else if (prepend) {
```
Comment (N-Wouda) [inline src/controllers/core/controller.js:62]: I had hoped you'd solve this one for me 👍 

We indeed needed a prepend option, and this is quite elegant. Good work.

File: src/controllers/historycontroller.js [hunk 8]
```diff
@@ -13,34 +13,26 @@ class HistoryController extends Controller
     {
         if (entry instanceof Customer)
         {
-            this.addCustomer(entry);
+            this._loadTemplate(
+                "src/views/template/history/customer.html",
+                "#history-events",
+                entry,
+                false,
```
Comment (N-Wouda) [inline src/controllers/historycontroller.js]: A shame Javascript doesn't have named parameters :(

File: src/controllers/historycontroller.js [hunk 9]
```diff
@@ -13,34 +13,26 @@ class HistoryController extends Controller
...
+                true
+            );
         }
 
         if (entry instanceof Container)
         {
-            this.addContainer(entry);
+            this._loadTemplate(
```
Comment (N-Wouda) [inline src/controllers/historycontroller.js]: No, don't add the call to `_loadTemplate` twice (i.e. in different `if`s). Set the loading string based on the `instanceof` check, and then have this piece of code in your code only once with that string.

File: src/views/template/history/container.html [hunk 10]
```diff
@@ -1,3 +1,3 @@
-<li class="list-group-item">
-    Container
-</li>
\ No newline at end of file
+<li class="list-group-item list-group-item-success">
+    {{#l}}New rack of size{{/l}} {{ capacity }} {{#l}}added{{/l}}.
```
Comment (N-Wouda) [inline src/views/template/history/container.html:2]: 👍

**Comment by N-Wouda** [comment]
>Might need to add states to order the Order model, to determine the right Bootstrap classes. But that would be better to have its own PR.

Agreed. This should probably be part of the 'make it look nice' set of use cases.

**Comment by N-Wouda** [comment]
A few comments, but overall, I think this is close to merging.

File: src/views/template/history/customer.html [hunk 11]
```diff
@@ -0,0 +1,13 @@
+<li class="list-group-item">
+    <div class="panel panel-default">
+        <div class="panel-heading">{{ name }}</div>
```
Comment (N-Wouda) [inline src/views/template/history/customer.html]: This `name` should also be wrapped in translation.

File: src/views/template/customerorder.html [hunk 12]
```diff
@@ -6,7 +6,7 @@
         <table class="table">
             <thead>
                 <tr>
-                    <th>{{#l}}Product{{/}}</th>
+                    <th>{{#l}}Product{{/l}}</th>
```
Comment (N-Wouda) [inline src/views/template/customerorder.html]: Well spotted!

File: src/views/template/history/customer.html (hunk 11 above)
Comment (lukko) [reply src/views/template/history/customer.html]: Are we going to translate a Dutch name directly to another English name?

File: src/controllers/historycontroller.js (hunk 9 above)
Comment (lukko) [reply src/controllers/historycontroller.js]: Is this okay?

```javascript
class HistoryController extends Controller
{
    view()
    {
        super._loadTemplate(
            "src/views/template/history.html",
            "#history",
            {}
        );
    }

    log(entry)
    {
        if (entry instanceof Customer)
        {
            this._loadTemplate('customer', entry);
        }

        if (entry instanceof Container)
        {
            this._loadTemplate('container', entry);
        }

        MODEL.history.push(entry);
    }

    _loadTemplate(location, entry)
    {
        super._loadTemplate(
            "src/views/template/history/" + location + ".html",
            "#history-events",
            entry,
            false,
            true
        );
    }
}

```

Comment (N-Wouda) [reply src/controllers/historycontroller.js]: Why not something like,

```javascript
    log(entry)
    {
        var loc = "";

        if (entry instanceof Customer) {
           loc = "customer";
        } else if (entry instanceof Container) {
            loc = "container";
        } else {
            return;
        }

        this._loadTemplate(
            "src/views/template/history/" + loc + ".html",
            "#history-events",
            entry,
            false,
            true
        );

        MODEL.history.push(entry);
    }
```

File: src/views/template/history/customer.html (hunk 11 above)
Comment (N-Wouda) [reply src/views/template/history/customer.html]: English names to Dutch equivalents. Or two sets of names. I'm not too sure, but something along those lines, yes.

File: src/controllers/historycontroller.js (hunk 9 above)
Comment (lukko) [reply src/controllers/historycontroller.js]: Haha, I had that before I wrote down the above version.
Will change it :)

**Comment by N-Wouda** [comment]
Looks good to me! I guess this can merged then?

**Pull request merged**